engine/common/stop/stop_control.go

```diff
@@ -0,0 +1,147 @@
+		AddWorker(sc.processEvents).
+		Build()
+	sc.Component = sc.cm
+
+	return sc
+}
+
+// OnVersionUpdate is called when a version update occurs.
+//
+// It updates the incompatible block height and the expected node version
+// based on the provided height and semver.
+//
+// Parameters:
+//   - height: The block height that is incompatible with the current node version.
+//   - version: The new semantic version object that is expected for compatibility.
+func (sc *StopControl) OnVersionUpdate(height uint64, version *semver.Version) {
+	// If the version was updated, store new version information
+	if version != nil {
+		sc.log.Info().
+			Uint64("height", height).
+			Str("semver", version.String()).
+			Msg("Received version update")
+
+		sc.versionData.Store(&VersionMetadata{
+			incompatibleBlockHeight: height,
+			updatedVersion:          version.String(),
+		})
+		return
+	}
+
+	// If semver is 0, but notification was received, this means that the version update was deleted.
+	sc.versionData.Store(nil)
+}
+
+// onProcessedBlock is called when a new block is processed block.
+// when the last compatible block is processed, the StopControl will cause the node to crash
+//
+// Parameters:
+//   - ctx: The context used to signal an irrecoverable error.
+func (sc *StopControl) onProcessedBlock(ctx irrecoverable.SignalerContext) {
+	versionData := sc.versionData.Load()
+	if versionData == nil {
+		return
+	}
+
+	newHeight := sc.lastProcessedHeight.Value()
+	if newHeight >= versionData.incompatibleBlockHeight-1 {
+		ctx.Throw(fmt.Errorf("processed block at height %d is incompatible with the current node version, please upgrade to version %s starting from block height %d",
+			newHeight, versionData.updatedVersion, versionData.incompatibleBlockHeight))
+	}
+}
+
+// updateProcessedHeight updates the last processed height and triggers notifications.
+//
+// Parameters:
+//   - height: The height of the latest processed block.
+func (sc *StopControl) updateProcessedHeight(height uint64) {
+	sc.processedHeightChannel <- height
+}
+
+// RegisterHeightRecorder registers an execution data height recorder with the StopControl.
+//
+// Parameters:
+//   - recorder: The execution data height recorder to register.
+func (sc *StopControl) RegisterHeightRecorder(recorder execution_data.ProcessedHeightRecorder) {
+	recorder.AddHeightUpdatesConsumer(sc.updateProcessedHeight)
+}
+
+// processEvents processes incoming events related to block heights and version updates.
+//
+// Parameters:
+//   - ctx: The context used to handle irrecoverable errors.
+//   - ready: A function to signal that the component is ready to start processing events.
+func (sc *StopControl) processEvents(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+	ready()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case height, ok := <-sc.processedHeightChannel:
+			if ok && sc.lastProcessedHeight.Set(height) {
+				sc.onProcessedBlock(ctx)
+			}
+		}
```

Comment:
```suggestion
		case height, ok := <-sc.processedHeightChannel:
			if !ok {
				return
			}
			if sc.lastProcessedHeight.Set(height) {
				sc.onProcessedBlock(ctx)
			}
		}
```

if `sc.processedHeightChannel` is closed we don't have to stay in this loop.